📖 [GEP-23] Autoscaling Shoot `kube-apiserver` via Independently Driven HPA and VPA (#7410)
gardener/gardener: docs/proposals/22-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

---
title: Autoscaling Shoot `kube-apiserver` via Independently Driven HPA and VPA  
gep-number: 0022
creation-date: 2023-01-31
status: implementable
authors:
- "@andrerun"
- "@voelzmo"

> reviewers:

Reviewer: Who will be the main reviewers for this GEP?

Author: Added oliver-goetz as main reviewer.

2. The compute cost of ShootKapi workloads constitutes a major part of Gardener's overall compute cost. Small relative
   waste in sizing ShootKapis results in considerable absolute cost.
3. Kube-apiserver does not scale well horizontally, in a sense that extra replicas come with substantial compute overhead.   

In short, ShootKapi needs to be scaled:
- Within a broad range
- With reasonable accuracy, relative to actual resource consumption

The implication is that ShootKapi requires both horizontal and vertical scaling. The existing implementation utilized 
to that end is [HVPA].

### Existing Solution
[HVPA] strives to achieve a fusion of HPA and VPA recommendations:
`HvpaRecommendedCapacity = k * HpaRecommendedCapacity + (1 - k) * VpaRecommendedCapacity`
The underlying goal is to allow HPA and VPA to be driven by the same CPU and memory usage metrics, and overcome the
HPA-VPA incompatibility known to exist in that mode.

### Deficiencies of the Existing Solution
HVPA strives to achieve a linear combination of HPA's and VPA's scaling policies:
`HvpaRecommendedCapacity = k * HpaRecommendedCapacity + (1 - k) * VpaRecommendedCapacity`.

> However there is a nonobvious obstacle before this intent: both HPA and VPA's policies are practically unavailable.

Reviewer 2: As a minor feedback, this part ("both HPA and VPA's policies are practically unavailable") was not so clear to me.

Author: I made the following attempt at making it a bit clearer (added the underlined part below):
>However, there is a nonobvious obstacle before this intent: both HPA and VPA's policies are practically unavailable <ins>- HVPA does not know how HPA/VPA would scale the workload, as function of circumstances</ins>. The only information available is "how would HPA/VPA change the existing workload right now".

I don't know if the above clarifies anything. Please don't hesitate to reopen conversations. I'm only closing them, as a tool to manage the comments I need to act upon.

> However there is a nonobvious obstacle before this intent: both HPA and VPA's policies are practically unavailable.

Reviewer 3: ```suggestion
However, there is a nonobvious obstacle before this intent: both HPA and VPA's policies are practically unavailable.
```

The only information available is "how would HPA/VPA change the existing workload right now". So, in place of the
original intent, HVPA resorts to:
`dHvpaRecommendedCapacity = k * dHpaRecommendedCapacity + (1 - k) * dVpaRecommendedCapacity` (the 'd' symbols stand for
"rate of change").

Had HPA and VPA recommendations been always deterministically available, up to date, and with absolute accuracy,
that would result in the originally intended behavior. Simply put, in a 3-D space where one coordinate is the real-world
utilization data, and the other two are the horizontal and vertical scale suggested by the recommender, this is a line -
a line which is a weighted average of the two lines which are individual HPA and VPA recommendations (Fig. 1).

> ![HVPA scaling](./assets/gep22-hvpa-scaling-space.png "Design outline")

Reviewer 2: ```suggestion
![HVPA scaling](./assets/gep22-hvpa-scaling-space.png "HVPA scaling space")
```

Author: Thanks for catching!

> _Fig. 1: HVPA policy as a weighted average of HPA and VPA policies. HPA and VPA policies in blue. HVPA intended policy

Reviewer 2: Maybe another sketch could be added that illustrates a problematic sequence of horizontal and vertical scaling. I was considering whether such a sketch with a single scaling sequence can help to understand this 3D diagram that shows the scaling space.

Author: A sketch of the extreme case where only scaling on one axis occurs, looks completely unremarkable - just a single straight line with intermediate points in it. And a more visually engaging scenario - one with some scaling on the other axis, and maybe even paradoxical downscaling (where VPA downscales immediately after HPA scales out), becomes verbose to explain and harder to understand, because it demonstrates multiple points at the same time.

I opted for adding the following explanation instead of a sketch (will be available in next push):
>As example of extreme single axis scaling, consider the following. A workload with constant memory consumption is being scaled by a HVPA configured at 50% HPA + 50% VPA, starting at 1 replica with 1 core. The expectation is that HVPA will maintain the number of replicas approximately equal to the number of cores per replica. Consider the following sequence of events, as CPU demand increases:
    1. CPU usage starts at 0.7 cores. HVPA is stable at 1 replica, 1 core
    2. CPU usage increases to 1.5 cores. HPA acts first, recommending 2 replicas, a change of +1. 50% of +1 is +0.5 replicas, rounded up to +1. HVPA applies +1 replica. The workload is now at consumption=1.5,Replicas=2,CoresPerReplica=1
    3. VPA acts, but does not recommend upscaling because CPU is currently over-provisioned. It does not recommend downscaling because MinAllowed is 1 core.
    4. CPU usage increases to 2.1 cores. HPA acts first, recommending 3 replicas, a change of +1. HVPA applies
       2 + ceil(50% * +1) = 3 replicas. The workload is now at Consumption=2.1,Replicas=3,CoresPerReplica=1
    ...and so on. HVPA acts as 100% HPA + 0% VPA, instead of the configured 50%/50%.

Author: Closing conversation to keep the PR manageable. Please reopen if necessary.

in green lies in the plane defined by HPA and VPA. In orange, a verical-only and

> horizontal-only scaling events, leading to deviation from intended policy._

Reviewer 3: ```suggestion
in green lies in the plane defined by HPA and VPA. In orange, a vertical-only and
horizontal-only scaling events, leading to deviation from the intended policy._
```

In practice though, while the HPA and VPA
abstract algorithms are deterministic, **when** they will be applied is not. At any point of the original "line", there
is a race between HPA and VPA, who will act first. If either HPA or VPA manages to act first, the line gets pulled only
in the direction of the horizontal or vertical scale axis. Not only is this new point not on the intended line, but this
new state feeds back into HPA and VPA's decisions, further influencing them. It is noteworthy that there is no force

> acting toward convergence to the original line, because once single-axis scaling occurs, it alleviates the pressure to

Reviewer 3: ```suggestion
acting toward convergence to the original line, because once a single-axis scaling occurs, it alleviates the pressure to
```

Author: Done. Changed to
>because once `a` single-axis scaling `event` occurs

scale on the other axis.

> To illustrate the above, if VPA is slow, it is possible for horizontal scale to reach its allowed maximum, with

Reviewer 3: ```suggestion
To illustrate the above, if VPA is slow, it is possible for the horizontal scale to reach its allowed maximum, with
```

vertical scaling never occurring.

In result, instead of a line in said space, the behavior of HVPA corresponds to an entire 2-D surface. For any value of
the actual workload resource consumption, there are multiple combinations of horizontal vs. vertical scale, which HVPA
would recommend, and the choice among these is accidental. Furthermore,
HPA and VPA do not act in accord, and when driven by the same input metrics, as is the case with
HVPA, they have different scaling trigger threshold levels. The mutual relation between these levels is entirely
accidental. Similarly, the resulting autoscaler lacks joint hysteresis behavior, and instead applies hysteresis
separately on the horizontal and vertical scaling axis, on arbitrarily aligned levels. 

The cumulative result is that HVPA's recommendations have intrinsic uncertainty between the choice of horizontal vs.

> vertical scaling, and a unique risk of oscillating on closed contours on aforementioned 2-D surface, circumventing

Reviewer 3: ```suggestion
vertical scaling, and a unique risk of oscillating on closed contours on the aforementioned 2-D surface, circumventing
```

HVPA is also organized around a concept of multiple scaling stages, where each stage has a different quotient for mixing
HPA and VPA recommendations, and the choice of stage is determined by the number of replicas. The intent is to achieve
a "horizontal first" or "vertical first" type of behavior. The proper functioning of this behavior relies on HVPA
decremental scaling returning on approximately the same trajectory, which is followed during incremental scaling.
However, due to the intrinsic uncertainty between the choice of horizontal vs. vertical scaling, this is not the case.
For example, workloads with "horizontal first" HVPA, which applies only HPA recommendations at low replica counts,
have been occasionally observed to grow to a large number of large replicas, then to reduce the number of replicas
before they have been downscaled vertically, and to end up in the low replica count, "horizontal first" stage, where
vertical scaling is not applied, with large replica(s), preserving large idle pods indefinitely.

As a mitigation of the horizontal vs. vertical scaling uncertainty, it is currently prescribed that HVPA only be used

> with weight coefficient (`k` in above formulae) of `100%` or `0%`. As far as HVPA core logic for fusing HPA and VPA,

Reviewer 3: ```suggestion
with weight coefficient (`k` in above formulae) of `100%` or `0%`. As far as HVPA core logic for fusing HPA and VPA is concerned,
```

that renders HVPA functionally equivalent to a much simpler implementation, reducing said core logic to nothing more
than unnecessary maintenance burden.

Kube-apiserver is currently scaled with HVPA strategy where the early stage is HPA-only and that stage results in

> excessive number of small replicas. VPA is unused until the number of replicas reaches specified maximum, and replicas

Reviewer 3: ```suggestion
an excessive number of small replicas. VPA is unused until the the number of replicas reaches the specified maximum, and replicas
```

retain their initial, minimal size. Kube-apiserver has low compute efficiency in that mode. As a mitigation, a

> `ScalingClass` heuristic estimate is applied, to predict the overall kube-apiserver load for the cluster, and set a

Reviewer 3: ```suggestion
`ScalingClass` heuristic estimate is applied, to predict the overall kube-apiserver load for the cluster and set a
```

better, but still fixed, initial replica size. However, while this substantially improves efficiency, it is still less

> efficient than using VPA, as it is impossible for a single, fixed value to match the compute resource needs of given

Reviewer 3: ```suggestion
efficient than using VPA, as it is impossible for a single, fixed value to match the compute resource needs of a given
```

vertical scaling until the recommendation differs from the current allocation (request) by said amount.
Alone, this concept is functionally incorrect when applied to scale-up: scaling is suppressed in a state where VPA is
indicating that the workload is starving for resources. To mitigate this problem, the existing design relies on
resource headroom existing on the node, which can temporarily accommodate the starving workload's needs. In the field,
this approach has been repeatedly observed to cause problems for etcd, for which a particularly high value of MinChange
is used, as frequent evictions are highly undesirable. As an illustration, take a pod with MinChange for CPU set to
1000 millicores, current CPU request at X millicores, actual pod need at X+2000, and available node headroom at 500.
What happens is that actual pod usage extends as far as headroom allows, reaching X+500 and is starving for extra 1500
millicores. Based on pod usage at X+500, VPA recommends e.g. X+900. This recommendation is below the MinChange threshold
of X+1000. HVPA does not act, and the pod is permanently stuck starving for 1500 millicores of CPU.

> Lastly, HVPA applies additive model to the interaction between HPA and VPA, with respect to provided compute power, but

Reviewer 3: ```suggestion
Lastly, HVPA applies an additive model to the interaction between HPA and VPA, with respect to provided computing power, but
```

that interaction is actually multiplicative. As an extreme example, take a workload running at one replica with one core.
Then consumption changes to 5 cores, leading HPA to recommend 5 single-core replicas - a change of +4 replicas. VPA
suggests adding 4 cores to the existing single replica. A HVPA configured at 50% HPA + 50% VPA, would apply half of
each recommendation, +2 cores and +2 replicas, resulting in (1 + 2) replicas * (1 + 2) cores. Failing to account for

> the multiplicative interaction between changes, HVPA provides a total of 9 cores to a workload which only needs 5. 

Reviewer 2: Overall, the examples and the whole section gives a good overview of the intricate deficiencies of the existing solution. Thank you for the detailed explanation! 😊

### Other Benefits to Replacing the Existing Solution
- HVPA applies recommendations via direct edits to the pod objects. This approach requires nonobvious special handling
  by Gardener code outside HVPA, which is error-prone.

### Goals
- Improve ShootKapi autoscaling stability by eliminating the stability issues existing in the current HVPA-based solution
- Improve ShootKapi operational efficiency at low application load levels, by avoiding excessive replicas driven by
  HVPA's policy for early horizontal scale-out
- Allow usage of VPA for right-sizing the resource requests during the entire time and not only under certain conditions
- Enable future migration away from HVPA by removing all dependency on it for scaling ShootKapi

> - Reduce Gardener custom code base and ongoing maintenance effort

Reviewer 3: ```suggestion
- Improve ShootKapi autoscaling stability by eliminating the stability issues existing in the current HVPA-based solution.
- Improve ShootKapi operational efficiency at low application load levels, by avoiding excessive replicas driven by
  HVPA's policy for early horizontal scale-out.
- Allow usage of VPA for right-sizing the resource requests during the entire time and not only under certain conditions.
- Enable future migration away from HVPA by removing all dependency on it for scaling ShootKapi.
- Reduce Gardener custom code base and ongoing maintenance effort.
```

### Non-Goals
- There are other components currently scaled via HVPA. Changing the autoscaling mode for these is outside the scope

>   of this GEP

Reviewer: Why? What's the plan with these other components? Will there be yet another scaling strategy for them? I think it makes sense to reduce complexity and also consider them as part of this GEP so that the goal of "dropping HVPA" can be achieved faster.

Reviewer 4: We're also planning to get rid of HVPA usage for the other components, but the approach will be entirely different. kube-apiserver is the only component using HVPA to combine VPA and HPA. All other components use HVPA to allow for tuning VPA configuration beyond what open-source VPA allows.
Therefore, we decided to keep this out of this KEP – it will be a separate KEP, covering a separate problem with a separate solution. Does this make sense?

Reviewer: Yes, it does. Can you please put exactly this explanation into the document then? I think this might definitely be of interested to other readers :)

Author: Done. Will be available with next push.

>   of this GEP

Reviewer 3: ```suggestion
- There are other components currently scaled via HVPA. Changing the autoscaling mode for these is outside the scope of this GEP.
```

- Changing the autoscaling of `kube-apiserver`s other than the one in the shoot control plane, is outside the scope of

>   this GEP

Reviewer 3: ```suggestion
- Changing the autoscaling of `kube-apiserver`s other than the one in the shoot control plane, is outside the scope of this GEP.
```

## Proposal
ShootKapi HVPA is replaced by individual HPA and VPA. Undesirable interference between HPA and VPA is avoided through

> use of sufficiently independent driving signals. HPA is driven by the rate of HTTP requests to kube-apiserver. VPA is

Reviewer 3: ```suggestion
the use of sufficiently independent driving signals. HPA is driven by the rate of HTTP requests to kube-apiserver. VPA is
```

driven by CPU and memory usage metrics.

### Rationale
A rough, heuristic horizontal sizing sets the stage for optimal VPA operation (horizontal size is within stable and
reasonably efficient range). Vertical scaling fine-tunes requests, ensuring high efficiency (requests not too high)

> and stability (requests not too low)

Reviewer 3: ```suggestion
and stability (requests not too low).
```

### Design Outline
One HPA and one VPA resources are deployed alongside each ShootKapi deployment object. The VPA controls CPU and memory
requests but not limits. The HPA is driven by the average rate of API requests per pod of the target kube-apiserver
deployment.

> An API request rate metric is deployed to shoot prometheus (one per shoot). It gets aggregated by the aggregate

Reviewer 5: As far as I'm aware, there is no monitoring stack deployed for clusters of purpose `testing` ([ref](https://github.com/gardener/gardener/blob/afe07938c905c47cef81565ebafc1c203dc96978/pkg/operation/operation.go#L576-L579)). But without a shoot prometheus, HPA scaling would not work. 

@istvanballok - would be great, if you could confirm/correct :) Thanks!

If that behavior is intended, it would be great to document it somewhere.

Reviewer 2: > As far as I'm aware, there is no monitoring stack deployed for clusters of purpose testing ([ref](https://github.com/gardener/gardener/blob/afe07938c905c47cef81565ebafc1c203dc96978/pkg/operation/operation.go#L576-L579)). But without a shoot prometheus, HPA scaling would not work.

Thank you for catching this aspect @hendrikKahl: it is correct that the monitoring stack is currently not deployed into the control plane of shoot clusters that are labeled with the testing purpose.

Author: Hi @hendrikKahl, and thanks for catching this! It is something we examined early in the design discussions, but then it fell out of my attention, as both the candidate approaches we settled upon, required negligible implementation effort.

The current plan is to have `purpose="testing"` clusters run in the "class-based scaling + HPA" mode, which is the currently supported alternative to HVPA.

The longer term vision is to support two scaling modes for shoot kube-apiserver:
1. The HPlusV autoscaling hereby proposed
2. Class-based vertical scaling + user enabled to set a fixed horizontal scale.

The above will be reflected in the GEP with the next push.

prometheus of the respective seed (one per seed) through Gardener's existing metrics aggregation mechanism.

A new component named prometheus-metrics-adapter is deployed on each seed. It sources performance metrics data from the
seed aggregate prometheus and makes it available in the K8s metrics format. The prometheus-metrics-adapter is registered

> as extension API service to the seed's kube-apiserver and takes responsibility of providing all custom metrics for that

Reviewer 3: ```suggestion
as an extension API service to the seed's kube-apiserver and takes responsibility of providing all custom metrics for that
```

kube-apiserver.

The request rate signal driving HPA originates in the shoot prometheus. It is propagated to the aggregate seed
prometheus, and then further mediated by the prometheus-metrics-adapter and exposed as custom metric in the seed

> kube-apiserver. The custom metric in turn is directly consumed by the HPA controller in the seed's garden namespace. 

Reviewer 2: > in the seed's garden namespace

I was not sure if this matches the diagram or what exactly it refers to.

Author: Good catch! That's a leftover from an early misconception I had - I started with the false idea that for seeds we deploy autoscaler controllers in the garden namespace. Also, there's the oversight that HPA does not run as a standalone pod (it's part of KCM, iirc).

![Design outline](./assets/gep22-design-outline.png "Design outline")

_Fig 2: Proposed design_

### Element: New Metric in Shoot Prometheus

> A new shoot:apiserver_request_total:sum recording rule is added to the shoot prometheus. To minimize latency, it

Reviewer 2: ```suggestion
A new `shoot:apiserver_request_total:sum` recording rule is added to the shoot prometheus. To minimize latency, it
```

> A new shoot:apiserver_request_total:sum recording rule is added to the shoot prometheus. To minimize latency, it

Reviewer 3: ```suggestion
A new `shoot:apiserver_request_total:sum` recording rule is added to the shoot prometheus. To minimize latency, it
```

calculates request rate across the minimum period for which two distinct samples are reliably available.

### Element: Metric Propagation from Shoot to Seed

> To propagate metric data from individual shoot control planes, to the seed level, the existing aggregate seed prometheus

Reviewer 2: ```suggestion
To propagate metric data from individual shoot control planes to the seed level, the existing aggregate prometheus in the seed's garden namespace
```

I suggest this minor rewrite to avoid confusion with the "[seed Prometheus](https://gardener.cloud/docs/gardener/monitoring/)" which is not involved here.

> To propagate metric data from individual shoot control planes, to the seed level, the existing aggregate seed prometheus

Reviewer 3: ```suggestion
To propagate metric data from individual shoot control planes to the seed level, the existing aggregate seed prometheus
```

is leveraged. The aggregate prometheus already has access to shoot prometheus data, and implements a wildcard-based
metric aggregation mechanism which, as long as certain naming convention is followed, aggregates the aforementioned new
shoot prometheus metric without any need for changes.

### Element: Prometheus Metrics Adapter
A new component, named prometheus-metrics-adapter is added to seed clusters. It is an instance of the
[kubernetes-sigs/prometheus-adapter]. It implements the
Kubernetes Custom Metric API, exposing prometheus data in the form of K8s metrics selectively - controlled by
configuration. A single replica is used, as highly available autoscaling setup is currently prevented by limitations in
the autoscalers utilized by Gardener.

### Element: Seed API Server Extension Service

> The prometheus-metrics-adapter is registered as API extension to the seed API server. It is deployed with minimal

Reviewer 3: ```suggestion
The prometheus-metrics-adapter is registered as an API extension to the seed API server. It is deployed with minimal
```

> deviations from default configuration, to reduce future merge effort. As a result, it currently takes charge not only

Reviewer 3: ```suggestion
deviations from the default configuration, to reduce future merge effort. As a result, it currently takes charge not only
```

ratio observed with the existing solution. It will be then gradually fine-tuned to the maximum value which does not
impact quality of service.

This proposal is the first stage of an incremental, two stage approach. It aims primarily at resolving the existing
stability issues inherent to HVPA and enabling (but not executing) scaling efficiency optimizations. The second stage
will be focused solely on further fine-tuning scaling efficiency to optimize ShootKapi compute cost over HVPA.
Details follow in the [Discussion and Limitations](#discussion-and-limitations) section.

### Element: VPA
A standard VPA operates on ShootKapi. Its primary purpose is to improve resource efficiency over what's possible via
HPA's coarse, replica-granularity scaling, by shrinking replicas down to match actual utilization.

This improves over the existing solution by having vertical scaling active the entire time and not only during certain
periods.

### Transition Strategy
The following steps will be executed over time, and provide a non-disruptive transition from the existing implementation
to the one hereby proposed.

> 1. Proposed scaling approach will be deployed behind a feature gate and disabled by default. Existing HVPA feature flag
>    will be preserved at this point, as its use is not limited to kube-apiserver scaling.
> 2. New feature gate gradually enabled until all ShootKapi workloads scaled by new approach (Note: this step applies to
>    individual garden/seed instances, and not to the code base as a whole)
> 3. At this point, the preexisting HVPA feature flag has no effects on ShootKapi scaling (Note: this step applies to
>    individual garden/seed instances, and not to the code base as a whole)
> 4. New feature flag promoted and eventually removed
> 5. (Outside the scope of this proposal) use of HVPA removed for other workloads and the preexisting HVPA flag removed

Reviewer 3: ```suggestion
1. Proposed scaling approach will be deployed behind a feature gate and disabled by default. Existing HVPA feature flag
   will be preserved at this point, as its use is not limited to kube-apiserver scaling.
2. New feature gate gradually enabled until all ShootKapi workloads scaled by new approach (Note: this step applies to
   individual garden/seed instances, and not to the code base as a whole).
3. At this point, the preexisting HVPA feature flag has no effects on ShootKapi scaling (Note: this step applies to
   individual garden/seed instances, and not to the code base as a whole).
4. New feature flag promoted and eventually removed.
5. (Outside the scope of this proposal) Use of HVPA removed for other workloads and the preexisting HVPA flag removed.
```

## Discussion and Limitations

> #### Prometheus-metrics-adapter precludes the use of sources of custom metrics

Reviewer 2: ```suggestion
#### Prometheus-metrics-adapter precludes the use of other sources of custom metrics
```

Author: Thanks!

With current K8s design, only one API extension can serve custom metrics. Per this GEP, the custom metrics extension

> point is occupied by prometheus-metrics-adapter.

Reviewer: Why do you mention this explicitly? Is this (only one API extension can serve custom metrics a problem or will it become one?

Author: It imposes constraint on future design choices. Currently, we have the freedom to register any custom metrics provider. After the change, if a need arises for custom metrics which cannot be provided by the prometheus-metrics-adaper, we'll be in a position where we need to figure out a way to provide them, without disrupting the metrics required by the proposed HPA.

Reviewer: Yeah, sure, but what is the alternative now? I don't think there is any since the approach relies on the custom metrics. Hence, we have to go down this path anyways, I assume?

Author: The main point is to keep this (occupying the custom metrics extension point) in mind as part of the cost of the proposed approach.

There are more specific concerns, but from the perspective of this review, I think they are not important enough to consider. For instance, an alternative implementation of the proposed approach is to have custom metrics provided by a small custom component, instead of prometheus-metrics-adapter. That alternative approach has the benefit that it could easily coexist with another custom metrics provider by selectively forwarding queries to it. But I think that's of negligible importance at present. It suffices to note that the proposed approach hinders future use of other sources of custom metrics.

Reviewer: Yes. Just for the record, maybe you can put this

> hat alternative approach has the benefit that it could easily coexist with another custom metrics provider by selectively forwarding queries to it.

into the document then. I doubt that we will ever need it, but just in case (and since we now discussed it here). Thanks.

Author: Done. Will be available in next push.

#### Heavyweight Prometheus infrastructure feeds signal back into cluster control loop

> The proposed signal flow brings two Prometheus instances (shoot prometheus and seed aggregate prometheus) into the shoot

Reviewer 2: ```suggestion
The proposed signal flow brings two Prometheus instances (shoot control-plane prometheus and aggregate prometheus) into the shoot
```

To avoid the confusion with the seed Prometheus

Author: Done and will be available with next push, but I'd like to explain, what I was driving at by using "seed aggregate prometheus":

Gardener's monitoring setup is complex enough so for those who are new to the project, or maybe even those who just do not follow developments closely, it is easy to lose track of its structure. I think for those people it's easier when prometheus instances are prefixed with indication whether they run as part of seed or shoot, e.g:
- seed aggregate prometheus
- seed cache prometheus
- seed main prometheus
- shoot control plane prometheus (And thanks for adding the "control plane" in your comment. I don't know if that was your intent, but it's exactly in the spirit of my concern. For someone who is in doubt whether "shoot prometheus" may be referring to one running inside the shoot, that's a welcome clarification)

I know that those aren't the correct names, and you are pointing me the correct ones. So I am not advocating that I use the incorrect ones in this PR. I just wanted to provide a monitoring-outsider perspective, because it may not be obvious to those who know the monitoring setup well.

cluster control loop. This has major implications on reliability. The alternative is a small, custom-built component
registered as seed custom metrics API extension, which directly calls the ShootKapi metrics endpoints and outputs
request rate.

The current approach was chosen solely because it leverages an existing open source component and avoids authoring a
custom one. The historical record of Gardener Prometheus components reliability is deemed acceptable for the purpose.  

The proposed complex signal path introduces additional latency to scaling response. At present, the resulting scaling
latency is deemed satisfactory.

While the scope of this GEP is limited to using the Prometheus signal path, the proposed design allows for an easy
substitution with a signal path based on aforementioned simple custom component alternative. This supports future
changes based on accumulated field experience, or evolving solution needs. 

> #### Observed fluctuations in actual compute efficiency affecting ShootKapi workloads

Reviewer 2: This section was not so clear to me so I was wondering if you could give an example for this? How did the reduction of effective compute power manifest itself?

Author: Sure. I'll create a separate discussion space for this.

requires a strictly incremental effort on top of this GEP.

An approximate 50% reduction (factor 0.5) of the effective compute power per unit of CPU was occasionally observed on
some productive ShootKapi instances. The change occurs with a quick cutoff transition, and has been observed for periods
or up to 2 hours. This variance currently forces substantial safety margin upon HPA's target metric value, effectively
necessitating more replicas to be deployed than required during usual operation.

The source is believed to be competition for compute resources with sibling workloads, either on the K8s, or
infrastructure level, but the circumstances have not been conductive to investigation after the fact.
A subsequent stage, outside the scope of this GEP is planned to focus on understanding the root cause, and further
optimizing HPA's target metric value accordingly.

Once this biggest source of inefficiency has been resolved, yet another round of scaling performance fine-tuning
is planned, where instead of overall request rate, HPA will be driven by a weighted average, reflecting the different

> compute cost of different request categories (e.g. a cluster-scoped LIST is more cpu intensive than a resource-scoped PUT)

Reviewer 6: Couldn't the different resource consumptions of request categories be another reason for a deadlock situation? 
Assuming there is a cluster which runs many resource intensive cluster-scoped list events with label selectors on tens of thousands config maps. In this case the maximum size of the ShootKapi instances (node capacity) could be reached, before the request rate is high enough to trigger HPA scaling up horizontally.
On the other side, if the request limit is too low and there are many "cheap" requests, HPA would scale up too early so that there are too many small ShootKapi instances which are not running in a resource efficient way.

Author: This precisely is the motivation behind moving to a weighted average in the subsequent optimisation stage.

With HPA driven by a signal which does not predict resource demand with perfect accuracy, there is a tradeoff between:
- minimising the overhead which is the horizontal safety margin of extra requested replicas (efficiency)
- risk of providing insufficient number of replicas under atypical load profiles (reliability)

In the end, I expect it will come down to simple cost-benefit optimisation. We're starting with conservative 
(low) requests/replica, and then field data will show how far we can optimise.

Author: Closing the conversation to keep the PR manageable. Please don't hesitate to reopen if anything else comes to mind.

> compute cost of different request categories (e.g. a cluster-scoped LIST is more cpu intensive than a resource-scoped PUT)

Reviewer 3: ```suggestion
compute cost of different request categories (e.g. a cluster-scoped LIST is more cpu intensive than a resource-scoped PUT).
```

> ## Alternatives

Reviewer 6: This might sound naive, but why not scaling up horizontally if there are too many `429 too many requests` errors  at the ShootKapi instances of a shoot?

Reviewer 4: This sounds like the correct thing to do: detect an overload situation as indicated by HTTP 429 responses and react to it by adding more instances. However, I think there are two drawbacks to this approach:
* If a single client (for whatever reason) hits only a single replica, causing the HTTP 429 responses, this approach will continue to scale up and the overload situation will still exist. I think if we're adding autoscaling in any way we have to ensure that our approach is guaranteed to improve the metric we're evaluating.
* There might be other indicators of overload as well, for example response times getting slower, HTTP 503 responses, or similar things. I think currently we are not sure that we understand the whole spectrum of how kube-apiserver indicates overload that we should react to by scaling up.

Author: At least to me, it does not sound naive, and I'd like to make use of the smarter algorithm built into API priority and fairness (and surfacing as `HTTP 429` errors).

To your question: In practice, because historically issues have been with resource exhaustion, and not request level throttling (or I'm oblivious of it). And in load tests we didn't see in-flight requests reaching anywhere near current limits. Also, since Gardener does not have API priority and fairness deployed yet, the exact behaviour of HTTP-429s is seen as something that's about to change.

In theory (and if we're looking at HTTP-429 as complete replacement of request count as driver), because HTTP-429s don't reflect memory starvation. Also, they provide a cue when to scale out, but not when to scale back in.

I think that once we have API priority and fairness, HTTP-429s could make a valuable supplement to the current metric, catching cases of CPU starvation which are missed by that metric, and potentially enabling us to use narrower safety margin.

Author: Marking this resolved to keep the PR manageable. Please don't hesitate to reopen, if you'd like to discuss further.

Below is a list of the most promising alternatives to the proposed approach. 

> 1. **Using VPA, plus a simple custom horizontal autoscaler which acts when VPA recommendation goes below or above a

Reviewer 2: I was thinking about elaborating on a similar approach a bit more.

Some considerations:

- the range of horizontal scaling for the ShootKapi is fairly limited. Due to the nature of the K8s control plane (the underlying database, etcd, can not be scaled horizontally) I think the maximum realistic number of ShootKapi replicas is much less than 100. Even in large clusters that are beyond the size where we feel confident, 4-5 ShootKapi instances are sufficient. To rephrase it differently, I think if there was a need to adding more replicas, other components could fail before that, like saturating disk I/O on etcd or the network bandwidth between etcd and the ShootKapi-s.
- so for non HA clusters, realistic values for the number of replicas could be 1,2,3. I'm not sure if we could reach 4 or 5, because for such large clusters the HA feature will probably be turned on.
- for HA clusters where the API servers are distributed to the 3 zones where the 3 etcd instances are running, we could realistically expect 3 or 6 ShootKapi replicas, so the range of choices is reduced to only 2 possibilities.

> the different compute cost of different request categories (e.g. a cluster-scoped LIST is more cpu intensive than a resource-scoped PUT)

I was considering this alternative approach because of the concern described below that the request rate might not be a good driving signal given the load differences that 2 different kinds of requests can incur.

Would it be possible to configure HPA based on the total memory *requests* with the following heuristic, assuming a max practicable target ShootKapi memory request per instance of e.g. 40GB, given typical 64GB nodes? The total memory request metric could follow a similar signal path as described in this GEP (cache Prometheus, Prometheus Metrics Adapter; no need for the shoot control plane and aggregate Prometheus in this path, because the pod Memory requests are available in the cache Prometheus via the kube-state-metrics exporter for all the shoot control planes in the seed)
- total memory requests (non HA case)
  - `> 40GB`, scale ->2
  - `> 80GB`, scale ->3
  - `> 120GB`, scale ->4
  - `< 100GB`, scale ->3 (hysteresis of 20GB)
  - `< 60GB`, scale ->2
  - `< 20GB`, scale -> 1

- total memory requests (HA case)
  - `> 120GB`, scale to 6
  - `< 100GB`, scale to 3
  
So basically I'm wondering if the HPA could be driven by memory (request) metrics instead of the request rate metric.

Note the implicit assumption I made here based on my experience that the ShootKapi instances are rather bound by memory than by CPU. Even instances that consume considerable amount of memory (>20GB) are not using a lot of CPU (<4 cores), so that given a typical node (16 cores, 64GB) we can assume that the if the memory request fits, the CPU request will fit as well, so that it might be sufficient to drive horizontal scaling solely by the memory requests.

Reviewer 6: Response to https://github.com/gardener/gardener/pull/7410#discussion_r1093037847

The sum of the memory usages might be a better metric than memory requests for this case.
When scaling up, the new replica will have the same memory request as the existing ones. Thus, it would immediately scale up again. However, the memory usage of the new instance should be less than the request.
For scaling down it is similar. 

While writing down these lines, I have the impression that this is pretty close to a scenario listed as limitation of VPA. 

>  Vertical Pod Autoscaler should not be used with the [Horizontal Pod Autoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#support-for-resource-metrics) (HPA) on CPU or memory at this moment. However, you can use VPA with [HPA on custom and external metrics](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#scaling-on-custom-metrics).

Author: We took this offline for a discussion over a broader communications channel. In the end, I guess it's valid to view the obstacle before driving HPA on memory requests/usage as a case of the officially declared HPA-VPA incompatibility, mentioned above by @oliver-goetz. We cannot control HPA/VPA hysteresis thresholds applied to the driving signal, so we cannot make them play nicely together.

Thanks to @istvanballok and @rickardsjp we had a chance for a broader discussion which touched on many aspects of autoscaling, and helped us realise the need to be outspoken in the GEP about the limitations imposed by the lack of proper multidimensional autoscaler.

   predetermined acceptable range:**
   Initial research indicates to be a promising solution. The solution described in this GEP was ultimately preferred
   because it does not require building custom components.   
2. **Adding the missing stabilization features to VPA via a custom recommender which is a minimal fork of the default 
   VPA recommender:**
   Not a standalone solution, but a potential part of any solution which leverages VPA but requires less disruptive
   scaling behavior. Research indicates to be a pragmatically viable solution, incurring minimal ongoing maintenance
   cost.

> Continued use of (an improved) HVPA does not show promise, due to the inherent algorithmic issues described above. 

Reviewer 2: ```suggestion
Continued use of (an improved) HVPA does not look promising, due to the inherent algorithmic issues described above. 
```

Author: Done. Will be available with next push.

## References
- [HVPA](https://github.com/gardener/hvpa-controller)

> - [kubernetes-sigs/prometheus-adapter](https://github.com/kubernetes-sigs/prometheus-adapter)

Reviewer 2: ```suggestion
- [HVPA]
- [kubernetes-sigs/prometheus-adapter]

[HVPA]: https://github.com/gardener/hvpa-controller
[kubernetes-sigs/prometheus-adapter]: https://github.com/kubernetes-sigs/prometheus-adapter
```

For the reference style link to work in line https://github.com/gardener/gardener/pull/7410/files#diff-ae525c0807c097429e7192d887d3fc15c07fcf4993ae033a9376bec9d217e05dR64,
we need to define the links following this syntax.

See https://daringfireball.net/projects/markdown/syntax#link

Author: Thanks!
IMO, even if we did not require it, it's still the better way to do the job, as it separates content from representation, so I'll gladly use it.

---

> title: Autoscaling Shoot `kube-apiserver` via Independently Driven HPA and VPA

Reviewer 7: The document should be linked in the docs index

Reviewer 7: This is not addressed

Author: Bumped document number to 23 and added to index.

> gep-number: 0022

Reviewer 7: ```suggestion
gep-number: 0023
```

Reviewer 7: This is not addressed

Author: I'll do this last, as it has no impact on reviews or implementation. I'd rather not introduce noise by repeatedly updating 22->23->24...

Author: Bumped document number to 23 and added to index.

      - [Gardener-custom metrics precludes the use of other sources of custom metrics](#gardener-custom-metrics-precludes-the-use-of-other-sources-of-custom-metrics)
      - [Observed fluctuations in actual compute efficiency affecting ShootKapi workloads](#observed-fluctuations-in-actual-compute-efficiency-affecting-shootkapi-workloads)
  - [Alternatives](#alternatives)
  - [References](#references)

## Summary
When it comes to autoscaling shoot control plane `kube-apiserver` instances (ShootKapi hereafter), Gardener needs
both stability and efficiency (accurate scaling). The existing approach of fusing HPA and VPA into the 2-dimensional
autoscaler which is HVPA, poses severe algorithmic limitations which manifest as stability and efficiency issues in the
field. 

This document outlines the main deficiencies in the existing solution and proposes a simple design which overcomes them,
reduces maintenance effort by reducing Gardener's custom code base, and enables subsequent reduction in ShootKapi
compute cost.

## Motivation
Gardener uses a relatively complex approach for scaling shoot control plane `kube-apiserver` instances (ShootKapi

> hereafter). This need is a consequence of the following concerns:

Reviewer 6: ```suggestion
Gardener uses a relatively complex approach for scaling shoot control plane `kube-apiserver` instances.
This need is a consequence of the following concerns:
```
Nit: `ShootKapi` is already introduced in `Summary`.

Author: Thanks! I intentionally duplicated the introduction, but didn't notice that it's annoying in that form (which it very much is).

I removed the "hereafter" from the second introduction. Now it reads just '(ShootKapi)' Hopefully that's a good compromise between the annoyance of a redundancy, and reinforcing the abbreviation in case it didn't stick in the reader's head the first time.

In short, ShootKapi needs to be scaled:
- Within a broad range
- With reasonable accuracy, relative to actual resource consumption

The implication is that ShootKapi requires both horizontal and vertical scaling. The existing implementation utilized 
to that end is [HVPA].

### Existing Solution
[HVPA] strives to achieve a fusion of HPA and VPA recommendations:
`HvpaRecommendedCapacity = k * HpaRecommendedCapacity + (1 - k) * VpaRecommendedCapacity`
The underlying goal is to allow HPA and VPA to be driven by the same CPU and memory usage metrics, and overcome the
HPA-VPA incompatibility known to exist in that mode.

### Deficiencies of the Existing Solution
HVPA strives to achieve a linear combination of HPA's and VPA's scaling policies:<br/>
`HvpaRecommendedCapacity = k * HpaRecommendedCapacity + (1 - k) * VpaRecommendedCapacity`<br/>
However, due to technical factors discussed below, it is not possible to fulfil this goal. At different
points in time, HVPA ends up mixing horizontal and vertical scaling in proportion which arbitrarily varies in the
entire range between horizontal-only scaling and vertical-only scaling.

> #### Algorithmic limitation: fusing the behavior of two blackbox systems 

Reviewer 7: I still think this section is way too abstract. I get the point and agree with it, but putting such theoretical work in a proposal document doesn't provide any benefit.
But I can ignore the section for now.

Author: I extracted the algorithmic issues section to a separate document and referenced it from the GEP.

the use of sufficiently independent driving signals. HPA is driven by the rate of HTTP requests to kube-apiserver. VPA is
driven by CPU and memory usage metrics.

### Rationale
A rough, heuristic horizontal sizing sets the stage for optimal VPA operation (horizontal size is within stable and
reasonably efficient range). Vertical scaling fine-tunes requests, ensuring high efficiency (requests not too high)
and stability (requests not too low).

### Design Outline
One HPA and one VPA resources are deployed alongside each ShootKapi deployment object. The VPA controls CPU and memory
requests but not limits. The HPA is driven by the average rate of API requests per pod of the target kube-apiserver
deployment.

A new component named gardener-custom-metrics is deployed on each seed. It directly scrapes metrics data from all
ShootKapi pods on the seed and derives custom metrics based on it. The component is registered as an extension API
service to the seed kube-apiserver. It occupies the custom metrics API extension point and is responsible for providing
all custom metrics for the seed kube-apiserver, including the one driving the ShootKapi HPA.

> ![Design outline](./assets/gep22-design-outline.png "Design outline")

Reviewer 7: Nit: there is no "Soil" in gardener

Author: I think at some point we added it to the glossary: https://gardener.cloud/docs/glossary/

Author: ...but indeed it is not actually necessary that the seed control plane be hosted by a soil cluster. I'll replace "soil" with something vague. I wish we had a good general term for this...

Author: I replaced "Soil" with "Seed control plane host". Let me know in case you come up with a better name for the diagram box.

_Fig 2: Proposed design_

### Element: Gardener Custom Metrics Provider Component 
A new component, named gardener-custom-metrics is added to seed clusters. It periodically scrapes the metrics endpoints
of all ShootKapi pods on the seed. It implements the K8s custom metrics API and provides K8s metrics specific to
Gardener, based on custom calculations. The proposed design can naturally be extended to multiple sources of
input data and a non-volatile cache for acquired data. However, in the scope of this proposal, the only data source
is the aforementioned metrics scrape, and the calculated values only need to be briefly cached in memory.

A single replica is deployed, as highly available autoscaling setup is currently prevented by limitations in

> the autoscalers utilized by Gardener.

Reviewer: What does this mean? Which limitations?

Author: VPA doesn't have HA support. Multiple instances can work only if they are configured to target different namespaces. Running multiple replicas of the gardener-custom-metrics would still result in a minuscule improvement of expected availability, but it's not worth the extra running cost, since it wouldn't suffice for a full HA setup, with VPA's single replica.

I could be more specific. Is the following an improvement (edit: if placed in the GEP document):
_A single replica of the gardener-custom-metrics is deployed. HA operation is not currently supported for VPA. That prevents an overall HA autoscaling setup and renders the use of multiple gardener-custom-metrics replicas economically unjustified._

Reviewer 7: I don't get this discussion.
We can deploy 2 replicas + PDB and it would work with VPA. I don't see an issue.

Author: VPA itself does not support leader election, so it's not possible to run it in K8s' typical active/passive HA arrangement. Since we don't get highly available VPA, we don't get nominally HA autoscaling. So it's not justified to pay the cost of extra gardener-custom-metrics replicas, when that would still not result in nominally HA autoscaling.

I'm saying "nominally", because autoscaling generally does not require fast reactions. The time required to spin up a new VPA controller pod is an order of magnitude less than delays introduced by rate limiting on a loaded cluster.

The only scenario in which I know that lack of VPA controller redundancy could have impact is:
- VPA controller is in AZ1
- AZ1 fails
- There is lasting capacity shortage in AZ2, so the VPA controller cannot be scheduled

Reviewer 7: I still don't get it. 
If VPA can't use leader-election so that we need to deploy a single VPA replica, why can't we still have the custom-metrics-server (which happens to be scaled by VPA) deployed in HA mode (two replicas)?

Author: We could have multiple replicas of gardener-custom-metrics. It's just not worth the expense (IMO), since it does not achieve nominally HA autoscaling functionality. Gardener-custom-metrics is very similar to VPA - it starts quickly and its absence does not immediately disrupt availability. So it's hard to justify paying for redundant replicas if you'd still not be able to declare HA.

Author: Nothing stops gardener-custom-metrics from running in a most reliable active/active replication mode, but cost. The extra replica(s) would have minimal contribution to risk reduction, but would impart substantial cross-AZ traffic.

A kube-apiserver's metrics response measures megabytes in size. To reduce the amount of network traffic, this proposal
utilizes compressed HTTP responses when scraping metrics. Outside the scope of this proposal, a future enhancement is
possible, in which a metrics filter component is added as a sidecar to each ShootKapi pod. Each such sidecar would
mirror the metrics endpoint of its respective ShootKapi, with the only addition of support for filtering based on an
HTTP request parameter (e.g. `GET /metrics?name=apiserver_request_total` would forward a `GET /metrics` request to the
ShootKapi, and then would reduce the response to only `apiserver_request_total` counters, before passing it on to the
caller). 

### Element: New Custom Pod Metric for ShootKapis

> A new `shoot:apiserver_request_total:rate_by_pod` pod custom metric is made available for each ShootKapi pod on the

Reviewer 7: I would call the custom metric something like `apiserver_request_total:rate`. It isn't limited to shoots as such and `by_pod` is implied by the nature of a pod custom metric.
But at the end, this is an implementation detail and it doesn't have to be agreed upon right now.

Author: Good point! After the transition to gardener-custom-metrics, the metric no longer lives in Prometheus, so it's just a custom pod metric and  the "_by_pod" suffix is pointless.

Fixed and will be available in next push.

ratio observed with the existing solution. It will be then gradually fine-tuned to the maximum value which does not
impact quality of service.

This proposal is the first stage of an incremental, two stage approach. It aims primarily at resolving the existing
stability issues inherent to HVPA and enabling (but not executing) scaling efficiency optimizations. The second stage
will be focused solely on further fine-tuning scaling efficiency to optimize ShootKapi compute cost over HVPA.
Details follow in the [Discussion and Limitations](#discussion-and-limitations) section.

### Element: VPA
A standard VPA operates on ShootKapi. Its primary purpose is to improve resource efficiency over what's possible via
HPA's coarse, replica-granularity scaling, by shrinking replicas down to match actual utilization.

This improves over the existing solution by having vertical scaling active the entire time and not only during certain
periods.

### Transition Strategy
The following steps will be executed over time, and provide a non-disruptive transition from the existing implementation
to the one hereby proposed.

> 1. Proposed scaling approach will be deployed behind a feature gate and disabled by default. Existing HVPA feature flag

Reviewer: For simplicity, I guess we can deploy the custom metrics component unconditionally and only control which scaling approach is used in the `kubeapiserver` component via the feature gate.

Author: The motivation is strictly economical, and only applies to HA seeds. Gardener-custom-metrics contributes to cross-AZ traffic (approx. 65KB/ShootKapi/minute), so you want if off if you're not using it.

Edit: But yes, that creates a common dependency on the feature gate, for both gardener-custom-metrics and the autoscaler-proper part. I'd prefer the ability for a multi-step implementation, where gardener-custom-metrics can be independently developed, verified, and deployed.

   will be preserved at this point, as its use is not limited to kube-apiserver scaling.
2. New feature gate gradually enabled until all ShootKapi workloads scaled by new approach (Note: this step applies to
   individual garden/seed instances, and not to the code base as a whole).
3. At this point, the preexisting HVPA feature flag has no effects on ShootKapi scaling (Note: this step applies to
   individual garden/seed instances, and not to the code base as a whole).
4. New feature flag promoted and eventually removed.
5. (Outside the scope of this proposal) Use of HVPA removed for other workloads and the preexisting HVPA flag removed.

### Vision Beyond This Enhancement
This proposal is part of a mid-term vision which enables the user to select an alternative ShootKapi scaling mode
for any shoot. Said mode prioritises reliability, and is known to work well even under extreme API request profiles.

The following approach is envisioned (discussion is limited to scaling ShootKapi):
- The default scaling mode is the one proposed by this GEP.

> - For any shoot, the user is enabled to select an alternative scaling mode. In this mode, the user specifies a fixed

Reviewer 7: I don't think we should allow users to interfere with the autoscaling method on a per-shoot basis.
Operators should make sure to gain confidence in the new mechanism and roll it out gradually (on a per-seed basis if desired).

Author: >I don't think we should allow users

Makes sense! We should not complicate user API in the absence of clear need. I changed "user" to "operator". Change will be available with next push.

>Operators should make sure to gain confidence in the new mechanism and roll it out gradually

That's exactly the plan. However, the new scaling method has known limitations and cases where it would predictably fail. So we need a reliable fallback. If some user has a use case where 90% of their Kapi requests are cluster-wide LIST (or get by label), I think it's better to have the option of operator setting a manual override, than being forced to say "we don't support your use case".

Author: >it's better to have the option of operator (...) override

@timebertt, if you agree with the above, I'll mark the conversation resolved.

Reviewer 7: I don't really agree. We should strive towards a solution that also works for the extreme cases. There shouldn't be the need to overwrite things on a per-shoot basis.

Reviewer 7: If operators notice problems with the new mechanism they should disable it on the entire seed/environment and we should improve the code.

Author: I think we don't have an effective solution which would be guaranteed to work, short of implementing a proper 2D autoscaler. As long as we scale horizontally based on an (efficient) application load heuristic, the possibility exists that the heuristic will be incorrect for some case.

Multi-dimensional autoscaling is still in the design/proposal phase: https://github.com/kubernetes/autoscaler/blob/master/multidimensional-pod-autoscaler/AEP.md

Edit: This GEP is aiming at a "good enough" interim solution, which improves Gardener's resource utilisation at the cost of some extreme cases possibly requiring manual adjustments.

Reviewer 7: It seems that there is no room for discussion, so let's resolve this conversation.

Reviewer 4: I don't think this is something we can even try to resolve on the abstract level of a GEP. I also think that we shouldn't built in things that we don't have evidence that we need it. So for now, I'd say let's drop the option to configure this on a per-Shoot-basis – regardless of _who_ is doing it (operator or Shoot owner, or someone else) – and wait for the cases where we really need this option.

For now, I think the only option we should build is to have the feature gate allowing to roll this out gradually one Seed after another.

Author: I hate to think about it like that, but we really don't have that many options. During design, we considered 7+ different scaling approaches. The one proposed here wasn't the best, functionally, but it was the best which fits within our resource constraints. We'll fine tune parameters and the signal driving HPA in the future, but other than that, algorithmic changes are off limits, unless we fork VPA/HPA.

Author: @oliver-goetz @timebertt @voelzmo 
The section discusses future vision, and is not meant to be part of this GEP's implementation. But to me, that was for reasons of optimal project execution, not because I doubt its usefulness.

Maybe I'm missing something. Here is my thinking:

_When the unsuitability of the new scaling approach becomes an issue for some customer cluster, that incident will be a here-and-now live service issue. Not a theoretical design objection. With the proposed fallback option, the time to accommodate the customer demand will be measured in minutes. On the other hand, updating Gardener design will take months._

_The parameter fine-tuning (requests/second per replica) will be governed by economy. We'll be optimising for the frequent case, intentionally leaving extreme outliers out. Taking into account the broad diversity of shoot loads supported by Gardener, I expect that there will be occasional outliers._

Anyway, I'm in the clear minority here, so it's impractical to drag the discussion any further. I'll remove that part of the document.

field. 

This document outlines the main deficiencies in the existing solution and proposes a simple design which overcomes them,
reduces maintenance effort by reducing Gardener's custom code base, and enables subsequent reduction in ShootKapi
compute cost.

## Motivation
Gardener uses a relatively complex approach for scaling shoot control plane `kube-apiserver` instances (ShootKapi
hereafter). This need is a consequence of the following concerns:
1. Gardener supplies K8s clusters of highly diverse compute scale and utilization - ranging from a couple of small
   practically idle nodes, to hundreds of heavily utilized giants. Furthermore, the ratio of 'cluster compute capacity'
   to 'ShootKapi load' substantially varies between different applications utilizing Gardener shoots. As a result,
   compute resources required by ShootKapis of different shoots vary by two orders of magnitude.
2. The compute cost of ShootKapi workloads constitutes a major part of Gardener's overall compute cost. Small relative
   waste in sizing ShootKapis results in considerable absolute cost.

> 3. Kube-apiserver does not scale well horizontally, in a sense that extra replicas come with substantial compute overhead.   

Reviewer 7: I don't think this belongs here and I don't think this is actually true. 
Horizontal scaling always comes with overhead (base load of instances). 
However, I don't see how this is relevant for gardener's autoscaling as we will take kube-apiserver as it is.

Author: A limited effort was made to test that with both synthetic and field tests. Also a review of related online publications. In field tests, when processing comparable request load, compared to a baseline of 1 kube-apiserver replica, 3 replicas resulted in 52-81% memory overhead, and 115-136% CPU overhead.

>Horizontal scaling always comes with overhead (base load of instances).

Certainly - the consideration is actually quantitative. The statement in the GEP is in qualitative form, strictly for brevity.

The overhead of horizontal kube-apiserver scaling was one of the most influential factors in selecting the scaling approach. Had the overhead of extra replicas been acceptable, the preferred approach would have been:
- Set vertical scale based on cluster node count, aiming at expectancy of, say, 5 replicas
- Then use only HPA to scale horizontally

Author: By the way, I'd be very interested to hear in case you had a different experience with compute overhead, when adding extra kube-apiserver replicas. Personally, I was surprised to see greater overhead on the CPU side (as opposed to memory). Still don't know what to make of it, but I haven't had a chance to investigate.

Author: Marking this discussion thread resolved to keep the PR manageable. Please don't hesitate to reopen as appropriate.

In short, ShootKapi needs to be scaled:
- Within a broad range
- With reasonable accuracy, relative to actual resource consumption

The implication is that ShootKapi requires both horizontal and vertical scaling. The existing implementation utilized 
to that end is [HVPA].

### Existing Solution

> [HVPA] strives to achieve a fusion of HPA and VPA recommendations:

Reviewer 7: Please note that `HVPA` is not enabled by default (e.g., we don't enable HVPA in our environments).
The alternative is no vertical autoscaling for kube-apiserver (static resource requests based on the scaling class annotation on the shoot: `alpha.kube-apiserver.scaling.shoot.gardener.cloud/class`).
This also contributes to the motivation of having a single robust solution for all cases.

Author: >HVPA is not enabled by default 

True... but in the GEP I'm taking the same view, as the one Rafael expresses [above](https://github.com/gardener/gardener/pull/7410#issuecomment-1433029953), as I feel that it reflects the true state of the HVPA feature gate. HVPA is off by default in the implementation, but it seems to me that we have long ago accepted that the default for kube-apiserver is HVPA. We just never updated the implementation.

In the proposal, the "class-based scaling plus HPA" still has a role - as the costlier but highly reliable alternative, which can be used to cover extreme cases.

Author: >we don't enable HVPA in our environments

Ah! At first, I misinterpreted that.

I now understand that the "scaling class + HPA" is the primary scaling mode for you? Do you find the scaling class request levels appropriate? With class-based scaling alone, we often find ourselves looking at some cluster and wishing that there was a VPA to downscale. But then again, HVPA is not much better in practice, with its horizontal-first policy.

HPA-VPA incompatibility known to exist in that mode.

### Deficiencies of the Existing Solution
HVPA strives to achieve a linear combination of HPA's and VPA's scaling policies:
`HvpaRecommendedCapacity = k * HpaRecommendedCapacity + (1 - k) * VpaRecommendedCapacity`.
However, there is a nonobvious obstacle before this intent: both HPA and VPA's policies are practically unavailable -
HVPA does not know how HPA/VPA would scale the workload, as function of circumstances. The only information available
is "how would HPA/VPA change the existing workload right now". So, in place of the original intent, HVPA resorts to:
`dHvpaRecommendedCapacity = k * dHpaRecommendedCapacity + (1 - k) * dVpaRecommendedCapacity` (the 'd' symbols stand for
"rate of change").

Had HPA and VPA recommendations been always deterministically available, up to date, and with absolute accuracy,
that would result in the originally intended behavior. Simply put, in a 3-D space where one coordinate is the real-world
utilization data, and the other two are the horizontal and vertical scale suggested by the recommender, this is a line -

> a line which is a weighted average of the two lines which are individual HPA and VPA recommendations (Fig. 1).

Reviewer 7: TBH, I don't understand a single sentence of this theoretical discussion.

Author: I've done many iterations on this section... I can explain reasonably simply the discrepancy in terms of superficial behaviour. But as soon as I get to motivating "why we can't just fix it, and are going for a new solution", there's a sharp step up in complexity.

Now that you made me look into it again, I wonder if the description would work better, if I dedicate the short first paragraph ("HVPA strives to achieve...") strictly to how if fails in terms of behaviour, instead of mixing in information about the underlying algorithm.

The discussion about the core algorithmic issue can then be placed in a separate section, so it can be easily skipped. Right now it is mixed with the concern which follows it ("HVPA is also organised around a concept of multiple scaling stages..."), which forces the user to read the algorithmic part, just so they can find where the next point starts.

I think I'll try that. I think it will only be a small improvement, but with no downsides, so it's still worth it.

Author: Marking this thread 'resolved' to keep the PR manageable, but don't hesitate to reopen if necessary. I broke down the section into several subsections, each with its own title. Hopefully makes it a bit easier to navigate and read selectively. Will be available with next push.

What happens is that actual pod usage extends as far as headroom allows, reaching X+500 and is starving for extra 1500
millicores. Based on pod usage at X+500, VPA recommends e.g. X+900. This recommendation is below the MinChange threshold
of X+1000. HVPA does not act, and the pod is permanently stuck starving for 1500 millicores of CPU.

Lastly, HVPA applies an additive model to the interaction between HPA and VPA, with respect to provided computing power, but
that interaction is actually multiplicative. As an extreme example, take a workload running at one replica with one core.
Then consumption changes to 5 cores, leading HPA to recommend 5 single-core replicas - a change of +4 replicas. VPA
suggests adding 4 cores to the existing single replica. A HVPA configured at 50% HPA + 50% VPA, would apply half of
each recommendation, +2 cores and +2 replicas, resulting in (1 + 2) replicas * (1 + 2) cores. Failing to account for
the multiplicative interaction between changes, HVPA provides a total of 9 cores to a workload which only needs 5. 

### Other Benefits to Replacing the Existing Solution

> - HVPA applies recommendations via direct edits to the pod objects. This approach requires nonobvious special handling

Reviewer 7: HVPA mutates the higher-level controllers: `Deployment`, `StatefulSet`, etc. 
This is one main drawback of HVPA's design: it always triggers a full rolling update of the workload, which brings unnecessary disturbance.

Author: Fixed. Changed to:
>HVPA applies recommendations via direct edits to the controlled object (e.g. `Deployment`).

Will be available in next push.

  by Gardener code outside HVPA, which is error-prone.

### Goals
- Improve ShootKapi autoscaling stability by eliminating the stability issues existing in the current HVPA-based solution.
- Improve ShootKapi operational efficiency at low application load levels, by avoiding excessive replicas driven by
  HVPA's policy for early horizontal scale-out.
- Allow usage of VPA for right-sizing the resource requests during the entire time and not only under certain conditions.
- Enable future migration away from HVPA by removing all dependency on it for scaling ShootKapi.
- Reduce Gardener custom code base and ongoing maintenance effort.

### Non-Goals

> - There are other components currently scaled via HVPA. Changing the autoscaling mode for these is outside the scope

Reviewer 7: Can you mention why this is out of scope?
I.e., the other components use HVPA just for vertical scaling but with more configuration options than the upstream VPA allows...
With this, it becomes clearer why these cases are different to autoscaling the API server in both dimensions

Reviewer 4: I think we had the same comment already earlier: https://github.com/gardener/gardener/pull/7410/files#diff-ae525c0807c097429e7192d887d3fc15c07fcf4993ae033a9376bec9d217e05d

Maybe this diff got lost somewhere along the way?

Author: Yes, oversight on my part. The modification is there, but I applied it to the second bullet (other Kapis), without noticing that it's also relevant to the first one (other components scaled via HVPA). Will fix.

Author: Done. I changed the "Non-goals" section to read:
>There are other components currently scaled via HVPA. Changing the autoscaling mode of these, and of `kube-apiserver`s other than the one in the shoot control plane, is outside the scope of this GEP. These workloads are subject to a different mode of autoscaling (e.g. vertical-only, where HVPA is only used to stabilize VPA behavior), and it is advantageous from project execution perspective to address them as a separate concern.

Author: Marking this discussion thread resolved to keep the PR manageable. Please don't hesitate to reopen as appropriate.

  of this GEP.

> - Changing the autoscaling of `kube-apiserver`s other than the one in the shoot control plane, is outside the scope of

Reviewer 7: What `kube-apiserver` instances other than in the shoot control plane do you have in mind? 😄

Author: I guess, the one in the garden. Possibly anything that might exist in any hypothetical supporting role clusters. It's mostly to disrupt a possible impression on the reader that "if it's kube-apiserver, it's guaranteed to be scaled that way".

Author: Marking this discussion thread resolved to keep the PR manageable. Please don't hesitate to reopen as appropriate.

  only used to stabilize VPA behavior), and it is advantageous from project execution perspective to address these as a 
  separate concern.

## Proposal
ShootKapi HVPA is replaced by individual HPA and VPA. Undesirable interference between HPA and VPA is avoided through
the use of sufficiently independent driving signals. HPA is driven by the rate of HTTP requests to kube-apiserver. VPA is
driven by CPU and memory usage metrics.

### Rationale
A rough, heuristic horizontal sizing sets the stage for optimal VPA operation (horizontal size is within stable and
reasonably efficient range). Vertical scaling fine-tunes requests, ensuring high efficiency (requests not too high)
and stability (requests not too low).

### Design Outline
One HPA and one VPA resources are deployed alongside each ShootKapi deployment object. The VPA controls CPU and memory
requests but not limits. The HPA is driven by the average rate of API requests per pod of the target kube-apiserver
deployment.

An API request rate metric is deployed to shoot prometheus (one per shoot). It gets aggregated by the aggregate

> prometheus of the respective seed (one per seed) through Gardener's existing metrics aggregation mechanism.

Reviewer 7: I don't think it's a good idea to route metrics via two prometheus instances. 
Shoot prometheus scrapes kube-apiserver with an interval of `1m` and the aggregate prometheus scrapes shoot prometheus with an interval of `1m`.
I.e., the signal that HPA receives for autoscaling might be delayed by up to `2m`. The actual autoscaling action might be delayed even further by the sync period of the HPA controller.

This doesn't sound like a sound signal path for autoscaling.
I suggest using a single central prometheus instance per seed that scrapes all relevant API server metrics of all shoots as the source for the custom metrics API.

Author: >This doesn't sound like a sound signal path for autoscaling.

I wouldn't go as far as "unsound", but similar concerns are expressed in the "Heavyweight Prometheus infrastructure feeds signal back into cluster control loop" section. Overall, I did not see the extra latency causing availability problems during tests.

Theoretically, it should be possible, so it's just that tests did not create the specific conditions where latency would surface. But I also expect that the frequency of it happening in real world use will be relatively low. I imagine it would require a sharp spike in request rate, at a point where the request count is just below the horizontal scaling threshold, and the request profile is heavily biased toward costlier requests (as to exhaust the horizontal safety margin).

It's perfectly possible, but then it's also possible to e.g. overwhelm VPA with a sharp load spike. Do you know of a client timeout which would be exceeded specifically due to the extra latency? That the latency would be crossing some threshold where some self-remediation action would be abandoned?

On a related note, deploying a separate prometheus to serve autoscaling seems prohibitively costly at this point. I am investigating into the option of using the cache prometheus for his.

A new component named prometheus-metrics-adapter is deployed on each seed. It sources performance metrics data from the
seed aggregate prometheus and makes it available in the K8s metrics format. The prometheus-metrics-adapter is registered
as an extension API service to the seed's kube-apiserver and takes responsibility of providing all custom metrics for
that kube-apiserver.

The request rate signal driving HPA originates in the shoot prometheus. It is propagated to the aggregate seed
prometheus, and then further mediated by the prometheus-metrics-adapter and exposed as custom metric in the seed
kube-apiserver. The custom metric in turn is directly consumed by the seed's HPA controller.

![Design outline](./assets/gep22-design-outline.png "Design outline")

_Fig 2: Proposed design_

### Element: New Metric in Shoot Prometheus

> A new `shoot:apiserver_request_total:sum` recording rule is added to the shoot prometheus. To minimize latency, it

Reviewer 7: HPA should be used with pod-specific custom metrics.
Can you explain what this recording rule does?

Author: Thanks for pointing this out. I believe I'd be better off using a ":rate_by_pod" suffix, instead of ":sum". The following is the change I made, and also answers your question:

### Element: New Metric in Shoot Prometheus
A new `shoot:apiserver_request_total:rate_by_pod` recording rule is added to the shoot prometheus. It is the rate of API requests per second, broken down by ShootKapi pod. To minimize latency, it calculates request rate across the minimum period for which two distinct samples are reliably available.

I'll double check the new name against convention.

calculates request rate across the minimum period for which two distinct samples are reliably available.

### Element: Metric Propagation from Shoot to Seed
To propagate metric data from individual shoot control planes to the seed level, the existing aggregate prometheus in
the seed's garden namespace
is leveraged. The aggregate prometheus already has access to shoot prometheus data, and implements a wildcard-based
metric aggregation mechanism which, as long as certain naming convention is followed, aggregates the aforementioned new

> shoot prometheus metric without any need for changes.

Reviewer 7: Only the specific API server metrics that are actually used for autoscaling should be aggregated to not overload the aggregate prometheus

Author: Sure. The "wildcard-based metric aggregation mechanism" is pre-existing. The proposal only adds the single `shoot:apiserver_request_total:sum` metric it actually uses.

### Element: Prometheus Metrics Adapter
A new component, named prometheus-metrics-adapter is added to seed clusters. It is an instance of the
[kubernetes-sigs/prometheus-adapter]. It implements the
Kubernetes Custom Metric API, exposing prometheus data in the form of K8s metrics selectively - controlled by
configuration. A single replica is used, as highly available autoscaling setup is currently prevented by limitations in

> the autoscalers utilized by Gardener.

Reviewer 7: I don't understand: why can't we have two replicas?
And why is this important for this GEP?

Author: Multiple VPA replicas don't play nicely, when acting on the same resources. Important, because Gardener aims for high availability, based on multiple component replicas. The note basically says we don't have that for the autoscaling setup.

Author: Marking this discussion thread resolved to keep the PR manageable. Please don't hesitate to reopen as appropriate.

### Element: Seed API Server Extension Service
The prometheus-metrics-adapter is registered as an API extension to the seed API server. It is deployed with minimal
deviations from the default configuration, to reduce future merge effort. As a result, it currently takes charge not only
of the custom metric API, but also of the external metric API, preventing the deployment of alternative handlers for

> these API extension points. 

Reviewer 7: Why is this relevant for this GEP?
We won't be using the external metrics API, so we shouldn't register it.

Author: We're using the default configuration available with the metrics adapter with minimum modifications, to minimise maintenance effort (merges). I see both advantages and disadvantages to it, but all were minor so I just opted for minimal intervention on my part.

The reason I mentioned it in the review, is specifically so I can get second opinion from reviewers. Do you see any specific scenarios where it creates a problem, or is it the general concern that it unnecessarily pollutes the API extensions space?

### Element: HPA
ShootKapi is scaled horizontally via HPA operating on the aforementioned average kube-apiserver request rate custom
metric. Scale-in is delayed by a stabilization window. This avoids unnecessary flapping, the effects of which would
be further amplified by VPA-driven evictions, and serves as a general precaution against HPA-VPA resonance,

> by ensuring that the two autoscalers operate with different frequency.

Reviewer 7: I'm skeptical that the request rate is the right signal for horizontal autoscaling.
There is no inherent limit to the rate of requests a single instance can handle.
E.g., a single instance can probably handle a high rate of non-mutating requests (think "cheap gets/lists") but it might not be able to cope with a comparably low rate of mutating requests.
Hence, the request rate is not a good indicator of how well an instance's capacity is being used.

A more obvious measurement of an instance's capacity is the number of inflight requests.
There is an inherent limit to how many requests can be processed concurrently by the API server (configured via flags/shoot spec).

Do you have any experiment results or basis for your decision that the request rate should be used?

Author: The choice of request rate as driving signal is, to a large extent, by exclusion. Inflight requests (IR) were the first criteria considered. We did some synthetic tests with ClusterLoader2 agains existing clusters, but did not see IR come close to the default limit (400), even under what was practically stress-level load for the clusters. 

The other thing is, IR are (I'm speculating) a good predictor of CPU load, as long as CPU is the bottleneck. If performance is constrained by e.g. downstream I/O, suddenly IR are making suggestions which have nothing to do with actual CPU demand. For instance, imagine a case where average KAPI request latency is 50ms. Relocate Etcd where it induces extra 50ms latency (same throughput, same everything else), and you suddenly have the IR metric recommending doubling the CPU resources, when actual CPU demand has not changed at all.

I do share your reservations, albeit with some potential differences. My design preferences are toward scaling CPU and memory, based on actual CPU and memory usage. The hope for the proposed approach is that it makes an appreciable improvement in the accuracy of allocated cpu/memory, while releasing us from the maintenance burden of HVPA. There are no illusions that the solution is ideal, or even that it is a step toward the ideal, on the implementation level. Reservations are mostly expressed through the "The current approach was chosen solely..." and "Overall, multidimensional autoscaling is a job which..." sections.

The scaling threshold value for HPA is initially set to a conservative (lower-biased) estimate of the requests/replica
ratio observed with the existing solution. It will be then gradually fine-tuned to the maximum value which does not
impact quality of service.

This proposal is the first stage of an incremental, two stage approach. It aims primarily at resolving the existing
stability issues inherent to HVPA and enabling (but not executing) scaling efficiency optimizations. The second stage
will be focused solely on further fine-tuning scaling efficiency to optimize ShootKapi compute cost over HVPA.
Details follow in the [Discussion and Limitations](#discussion-and-limitations) section.

### Element: VPA
A standard VPA operates on ShootKapi. Its primary purpose is to improve resource efficiency over what's possible via
HPA's coarse, replica-granularity scaling, by shrinking replicas down to match actual utilization.

This improves over the existing solution by having vertical scaling active the entire time and not only during certain

> periods.

Reviewer 7: Also, only pods are scaled individually instead of rolling deployments. This is another significant advantage over HVPA

Author: Why significant? Hopefully soon it will become significant, after VPA starts applying updates in-place, without evicting pods. But right now, isn't it a relatively minor inconvenience (e.g. forcing Gardener to implement carry-over of resource values when reconciling Gardener components)?

### Transition Strategy
The following steps will be executed over time, and provide a non-disruptive transition from the existing implementation
to the one hereby proposed.

1. Proposed scaling approach will be deployed behind a feature gate and disabled by default. Existing HVPA feature flag
   will be preserved at this point, as its use is not limited to kube-apiserver scaling.
2. New feature gate gradually enabled until all ShootKapi workloads scaled by new approach (Note: this step applies to

>    individual garden/seed instances, and not to the code base as a whole).

Reviewer 7: What is meant by "gradually enabled"?

Author: "Gradually", only in the sense that it allows minimising risk by initially applying on a smaller number of workloads. In practice, it is a reference to the ability (and intent) to roll out on a seed-by-seed basis.

3. At this point, the preexisting HVPA feature flag has no effects on ShootKapi scaling (Note: this step applies to
   individual garden/seed instances, and not to the code base as a whole).

> 4. New feature flag promoted and eventually removed.

Reviewer 7: What are the criteria for promoting the feature gate? What steps need to be performed?

I could imagine, that proper monitoring needs to be introduced that allows developers/operators to judge the quality of the new solution.

Reviewer 4: I'm all for hard criteria for feature gate promotion, which concrete ideas do you have in mind to judge the quality of an autoscaling solution? 

Right now, I can come up with an idea about _efficiency_: how much of the requested resources of each instance is actually used, to ensure the resource requests are appropriately set for each instance. Understanding if the number of instances is set correctly seems to be harder, I don't know if there's a good metric other than "fewer instances is better".
Looking at _effectiveness_, this seems even harder to measure. We can watch out for an increase in rejected requests or increase in request latencies on the apiservers – but this will most likely be a case-by-case thing and seems hard to be done programmatically for all apiservers? Maybe there are other, better ways to understand the quality of an autoscaling solution?

Author: I have no specific formal criteria in mind, at this time. That would make the plan the same as with most changes:
- Appoint key metrics/tests (in this case, e..g. API availability and latency, resource allocations, resource consumption, allocation-consumption relation, etc.)
- Measure on a representative sampling set
- Explore results with focus on application fitness/QoS and resource-level efficiency.

@voelzmo has more experience with Gardener practices and may be aware of a need for a more formalised approach so he may want to correct/add something.

5. (Outside the scope of this proposal) Use of HVPA removed for other workloads and the preexisting HVPA flag removed.

### Scaling Clusters with No Monitoring Stack
Monitoring stack is not deployed on clusters with `testing` purpose. An alternative mode of autoscaling will be used for
such clusters: In the short term, they will be scaled via the existing "class-based scaling + HPA" mode, which is the

> current alternative to HVPA.

Reviewer 7: This relates to one of my comments above:

In addition to the `testing` purpose, it's possible to disable the shoot monitoring instances entirely via the gardenlet configuration. 
E.g., we don't use gardener's monitoring stack for shoots. Hence, we won't be able to make use of this feature.
This is another point for using a central prometheus instance per seed.

Author: On a related, but separate note:
Assume this GEP supported an extensibility option which enabled you to deploy your own provider at the "custom metrics" API extension  point. This provider would need to implement the necessary requests-per-second-per-kapi-pod-per-shoot metric, to replace the provider based on prometheus-metrics-adapter. Would that make the proposed scaling mode useful to you?

Author: >This is another point for using a central prometheus instance per seed.

I looked further into this and consulted the monitoring experts. The existing cache prometheus is the obvious candidate. However, having it authenticate against the individual ShootKapis would be a major technical complication. The alternative of adding additional (4th) prometheus instance on the seed, has its obvious downsides too. Overall, it is not clear that the benefits are worth it at this time (future changes could change the parameters of that calculation, I assume).

Author: This discussion is now outdated, due to adding the gardener-custom-metrics, and no longer depending on prometheus as data source.

### Vision Beyond This Enhancement
This proposal is part of a mid-term vision which:
1. Utilizes a different approach for scaling ShootKapi on `testing` clusters, to account for the absence of monitoring
   stack on them.
2. For any shoot, enables the user to select an alternative ShootKapi scaling mode which maximises reliability and is

>    known to work well even under extreme API request profiles.

Reviewer 7: Our goal should be to introduce a solution that works well across the board.
Mentioning in this GEP that another mode will be introduced later on that is better for certain cases, questions the feasibility of the proposal for me.

Additionally, users should not be able to configure such details as API server requests/limits. This increases the API surface unnecessarily and allows them to intervene too deeply regarding seed resource usage.

Reviewer 4: > Additionally, users should not be able to configure such details as API server requests/limits. This increases the API surface unnecessarily and allows them to intervene too deeply regarding seed resource usage.

I was absolutely arguing in the same direction, until I found out that this is what happens already today, but with worse means. Currently everything needs to be done by operators, because Cluster owners can't do this themselves. And currently we have to take measures to prevent the next reconciliation overwriting these changes.

> Our goal should be to introduce a solution that works well across the board.

In an ideal world, you're right. However, no matter how good your autoscaling approach is, it will always be limited by the size of an individual node and the mechanism preventing that the vertical scale of an apiserver will be so large that it no longer fits on a node: we introduced a `maxAllowed` value for all critical items a while back.
This is by no means the best solution, and we could think about complicating this a bit more by trying to find out reasonable `maxAllowed` values dynamically based on the used worker sizes – but currently we don't do this and I'm not sure if there's a good way to do this.
So for now, the only thing that you can do in these cases is manual intervention: choose bigger controlplane machines, put the apiservers on them, adapt the `maxAllowed` and set manual values for requests/limits/replicas.

Not offering this as a first-class option doesn't prevent it from happening – it only makes these cases more cumbersome to handle for everyone.

Author: I think that @voelzmo's comment covers the most important points. I only wanted to add that:
>will be introduced later

No, not later. Not in the sense you're using it. The currently existing fallback already has the property "maximises reliability and is known to work well even under extreme API request profiles". The part of the GEP to which you are referring, says that the new approach will have this property, not that it will be the first approach to have this property.

the shoot cluster control loop. This has major implications on reliability. The alternative is a small, custom-built
component registered as seed custom metrics API extension, which directly calls the ShootKapi metrics endpoints and
outputs request rate.

The current approach was chosen solely because it leverages an existing open source component and avoids authoring a
custom one. The historical record of Gardener Prometheus components reliability is deemed acceptable for the purpose.  

The proposed complex signal path introduces additional latency to scaling response. At present, the resulting scaling
latency is deemed satisfactory.

While the scope of this GEP is limited to using the Prometheus signal path, the proposed design allows for an easy
substitution with a signal path based on the aforementioned simple custom component. Such custom-built component,
also has the advantage that it could easily coexist with other metrics providers by selectively forwarding queries to
them. This supports future changes based on accumulated field experience, or evolving solution needs. 

> #### Observed fluctuations in actual compute efficiency affecting ShootKapi workloads

Reviewer 7: This section seems to be too detailed/confusing for an implementation proposal.

Author: Another reviewer requested that it be added. I don't have a strong opinion either way, but I guess it would be easier to converge toward consensus on the grounds "leave it in, those who don't need it can ignore it".

There is a known source of horizontal scaling inefficiency, which presents an opportunity for substantial further
improvement. It is not addressed by this proposal and instead deferred to a separate implementation stage, as it
requires a strictly incremental effort on top of this GEP.

An approximate 50% reduction (factor 0.5) of the effective compute power per unit of CPU was occasionally observed on
some productive ShootKapi instances. The change occurs with a quick cutoff transition, and has been observed for periods
or up to 2 hours. This variance currently forces substantial safety margin upon HPA's target metric value, effectively

> necessitating more replicas to be deployed than required during usual operation.

Reviewer 7: I don't understand this section. What is the problem?

Author: Periods were observed, when kube-apiserver took twice the time to process the same requests, without a known change in relevant conditions. Cause needs to be understood, as this variation interferes with estimating how many 'requests per second' a Kapi instance can process.

Author: Marking the conversation "closed" to keep the PR manageable. Please reopen if necessary.
